kube-proxy has to clear NodePort stale UDP entries

aojea opened the pull request:
**What type of PR is this?**

/kind bug
/kind failing-test
/kind flake

**What this PR does / why we need it**:

kube-proxy needs to clean the stale UDP conntrack entries for  NodePorts, however,
this was done before the iptables rules were programmed and when the NodePort service was added,
without having into account if endpoints were available.

The conntrack entries have to be cleared after the iptables rules are programmed and when
the service has endpoints available.

**Which issue(s) this PR fixes**:
Fixes #91236

**Special notes for your reviewer**:

I´ve found out that the flake failed constantly, if the nodeport port never cleared the conntrack entries (UDP) on the opened socket 
https://github.com/kubernetes/kubernetes/pull/98130#discussion_r562508326

We´ve modifed the test to connect in parallel to the ClusterIP, so we can verify that the pods and the services are working correctly, the connection to the ClusterIP works, the conntrack entry is as follows
```
secctx=system_u:object_r:unlabeled_t:s0 use=1\nudp      17 115 src=10.131.0.73 dst=10.128.2.75 sport=80 dport=12345 src=10.128.2.75 dst=10.131.0.73 sport=12345 dport=80 [ASSURED] mark=0
 no conntrack entry for the nodeport connection (edited) 
```

However, the NodePort connection doesn´t work, and conntrack entry is the following:
```
secctx=system_u:object_r:unlabeled_t:s0 use=1\nudp      17 28 src=10.128.2.74 dst=10.0.32.3 sport=12345 dport=30801 [UNREPLIED] src=10.0.32.3 dst=10.0.32.2 sport=30801 dport=12345 mark=0
```
the iptables entries are correctly installed, but the packet is not forwarded because is hitting that conntrack entry,

**Does this PR introduce a user-facing change?**:

```release-note
kube-proxy: fix a bug on UDP NodePort Services where stale conntrack entries may blackhole the traffic directed to the NodePort.
```

aojea commented:
/assign @thockin @danwinship 
/cc @JornShen 
/sig network

JornShen commented:
ipvs mode also need modify? @aojea

aojea commented:
> ipvs mode also need modify? @aojea

/hold
yep, thanks
In the meantime let´s see what Dan and Tim think about this change

thockin commented:
OK, let's have a real chat about this.

First, I thought that we would not add the nodeport rule if we couldn't open the local port, but that just doesn't seem to be true.

What I am left wondering is - why do we even bother with this?  If we can't open it, we just log it and move on.  In that case, the very thing we are trying to prevent (colliding with an app in the root netns) is not prevented.   We know that apps explicitly using ports in this range are rare.  So is this really even worth doing? 

What if we just took it all out?  Does anyone think this is really worth the effort any more?  At the time I thought it was very clever, but I no longer believe it's worth it's on weight.

@aojea - what do you think?  Leave the conntrack flush, I think, but all the rest could go, including e2e.

I was going to make the same argument for hostPort - and maybe I will.  Kubelet already doesn't re-open those if it restarts, so it's at best "sloppy".

aojea commented:
> @aojea - what do you think? Leave the conntrack flush, I think, but all the rest could go, including e2e.

the conntrack flush has to be there, without it the test for UDP NodePorts always fail, at least is what we did observe in https://github.com/kubernetes/kubernetes/pull/98130#discussion_r562508326 

> First, I thought that we would not add the nodeport rule if we couldn't open the local port, but that just doesn't seem to be true.

indeed, this surprised me too, but thinking more about that, you can't know in the apiserver that the port is free on the nodes, so it will be weird for an user to be able to create a NodePort and then that kube-proxy refused to create it.

> I was going to make the same argument for hostPort - and maybe I will. Kubelet already doesn't re-open those if it restarts, so it's at best "sloppy".


I had this discussion with @danwinship , and the conclusion I took from that conversation is that this is needed due to security, but iptables rules take precedence over the application using the port, so the traffic will be forwarded to the NodePort/HostPort  and the application will be "hijacked" by the Service. 
I think that this is the security problem that the port binding tries to mitigate, but as you say is not true, because we only log and install the rules anyway ... but I have the feeling that I'm missing something , maybe Dan can explain more,

thockin commented:
> you can't know in the apiserver that the port is free on the nodes, so it
will be weird for an user to be able to create a NodePort and then that
kube-proxy refused to create it

Exactly.  At some point you have to ask who is in charge.  In a kube
cluster, I think the kube API is in charge.  So really, the best we can do
here is warn.  Maybe we could warn more loudly (events?) but I don't think
we're willing to NOT claim a nodeport if a service says to.

The initial objective was security (ish).  "If there's something unexpected
on that port, don't hijack it".  Then it became "...and prevent anyone from
trying to use it in the future".  The assumption there is that we won't
hijack it.  But we do.  And we know nobody really looks at the logs.  So
all this logic is effectively pointless (IMO).  If the goal was simply to
log something if a port is already held, all we'd really need to do is try
to claim it, log if error, else close it.

If there was a simpler way to hold ports open, I might be more supportive
of this, but really NodePorts in the 30K range seem like a user error.  Pod
hostports are the bigger liability and frankly, are almost as bad a feature
as externalIPs.

So would it be net simpler, and just as effective, to do less work here?

On Sat, Jan 23, 2021 at 4:50 AM Antonio Ojea <notifications@github.com>
wrote:

> @aojea <https://github.com/aojea> - what do you think? Leave the
> conntrack flush, I think, but all the rest could go, including e2e.
>
> the conntrack flush has to be there, without it the test for UDP NodePorts
> always fail, at least is what we did observe in #98130 (comment)
> <https://github.com/kubernetes/kubernetes/pull/98130#discussion_r562508326>
>
> First, I thought that we would not add the nodeport rule if we couldn't
> open the local port, but that just doesn't seem to be true.
>
> indeed, this surprised me too, but thinking more about that, you can't
> know in the apiserver that the port is free on the nodes, so it will be
> weird for an user to be able to create a NodePort and then that kube-proxy
> refused to create it.
>
> I was going to make the same argument for hostPort - and maybe I will.
> Kubelet already doesn't re-open those if it restarts, so it's at best
> "sloppy".
>
> I had this discussion with @danwinship <https://github.com/danwinship> ,
> and the conclusion I took from that conversation is that this is needed due
> to security, but iptables rules take precedence over the application using
> the port, so the traffic will be forwarded to the NodePort/HostPort and the
> application will be "hijacked" by the Service.
> I think that this is the security problem that the port binding tries to
> mitigate, but as you say is not true, because we only log and install the
> rules anyway ... but I have the feeling that I'm missing something , maybe
> Dan can explain more,
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/98305#issuecomment-765980637>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVHPCPEYOKO2OFHBFETS3LAYBANCNFSM4WOJ4ENA>
> .
>

JornShen commented:
>will be weird for an user to be able to create a NodePort and then that
kube-proxy refused to create it

agree with @aojea 

I think we can add event report local port failed opened. @thockin
Because at least we can let user know what happen in the node. And seems external ip do the same

https://github.com/kubernetes/kubernetes/blob/fb276ce2015a8bab93dd2d6ff96ca23a657df67b/pkg/proxy/iptables/proxier.go#L1113-L1127 

And 
> a failure on programming the iptables rules, still
keep the socket open, also that traffic to the nodeport is consumed

seems not right, the kube-proxy will reclose it. 

https://github.com/kubernetes/kubernetes/blob/fb276ce2015a8bab93dd2d6ff96ca23a657df67b/pkg/proxy/iptables/proxier.go#L1600-L1609

look at line 1607. So here we adjust the order of open the port seems make no much difference.

if I understood fault, plz point out @aojea :)

danwinship commented:
> We know that apps explicitly using ports in this range are rare. So is this really even worth doing?

Were nodeports always limited to a range? Or did that come later and so the port opener used to make more sense?

> I was going to make the same argument for hostPort - and maybe I will.

Do we limit hostports to a specific range? We need to prevent people from using hostports to snipe nodeports (and system services) (and hostnetwork pods).

> At some point you have to ask who is in charge. In a kube cluster, I think the kube API is in charge.

So that's interesting. In OpenShift 4.x, we have pretty much declared that you are not allowed to run your own non-containerized software on an OpenShift node, but I had assumed that people didn't feel that way about kubernetes clusters in general. Eg, kubernetes explicitly takes into account the fact that another piece of software on the system may accidentally delete its iptables rules out from under it; and there was some discussion in #92938 about whether it's reasonable for kube-proxy to change the value of host-wide kernel networking sysctls.

If the correct response to "I installed software behind kubernetes's back and then kubernetes stole its port away from it" is "Don't do that then!" then yeah, the port opener seems a lot less necessary.

(And maybe there could be some way to explicitly indicate to kubernetes that, eg, you're using port 22 for something, so stay away?)

aojea commented:
> If the correct response to "I installed software behind kubernetes's back and then kubernetes stole its port away from it" is "Don't do that then!" then yeah, the port opener seems a lot less necessary.

but why holding the port open? do we want to prevent people accidentally opening services behind forwarded ports?

JornShen commented:
> > If the correct response to "I installed software behind kubernetes's back and then kubernetes stole its port away from it" is "Don't do that then!" then yeah, the port opener seems a lot less necessary.
> 
> but why holding the port open? do we want to prevent people accidentally opening services behind forwarded ports?

@aojea 

look at the notes
https://github.com/kubernetes/kubernetes/blob/b3dd01da497f4f9f81ffb2ebf275ff08cbfebaa2/pkg/proxy/iptables/proxier.go#L1099-L1108

I think opening the port is a protective measures.

aojea commented:
> I think opening the port is a protective measures.

yeah, you are right, but that is for ExternalIP that also has its own CVE :smile: . 

What about make the port holding configurable, something like Selinux:
- Enforce: it keeps the socket open (not install the rule if it can not open it)
- Permissive: try to open a socket to check if there is something and log (current behavior) or just log and don´t hold the soket open (I prefer this and set it as default)
- Diabled: not try to hold at all

aojea commented:
ok, my hunch was right but I was wrong about the reason, it is not the socket thing, it is the the conntrack cleaning who was to be done after the rules are programmed, or it can happen that during the time that we clean conntrack for the UDP NodePort and the rules are programmed a flow is inserted. This flows goes stale and is renewed with the new connections, hence the test always fails to reach the first pod.

I just need to confirm that this actually fixes the problem, but  we have a way to reproduce it now, so I will test it tomorrow

This PR. moves the code to clean conntrack after the iptables rules are programmed, so it should solve the issue too.

cc: @smarterclayton

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1617,6 +1593,31 @@ func (proxier *Proxier) syncProxyRules() {
 	}
 
+	// For ports on node IPs, open the actual port and hold it.
+	for _, lp := range localPorts {
```
Note to self, check L1583
```
		utilproxy.RevertPorts(replacementPortsMap, proxier.portsMap)
```

smarterclayton commented:
> If the correct response to "I installed software behind kubernetes's back and then kubernetes stole its port away from it" is "Don't do that then!" then yeah, the port opener seems a lot less necessary.

Historically this is the logic I would have applied for node ports inside the range (by giving a range you assign the range to kube).  If someone can pick the port outside the range that becomes a different story (service steals 22), but if you are allowing end users to take arbitrary ports you already have the problem of (ssh crashed and kube gave 22 to bad_gal) regardless of port opener without a higher level control point preventing you accessing ports arbitrarily.  Someone running services on a node in the node port range is asking for trouble, like someone running local iptables mappings for the service cidr.

Commit:
kube-proxy: clear conntrack entries after rules are in place

Clear conntrack entries for UDP NodePorts,
this has to be done AFTER the iptables rules are programmed.
It can happen that traffic to the NodePort hits the host before
the iptables rules are programmed this will create an stale entry
in conntrack that will blackhole the traffic, so we need to
clear it ONLY when the service has endpoints.

aojea commented:
/retest

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1623,6 +1613,25 @@ func (proxier *Proxier) syncProxyRules() {
+
+	// Clear conntrack entries for the new UDP NodePorts, this has to be done AFTER the iptables rules are programmed.
+	// It can happen that traffic to the NodePort hits the host before the iptables rules are programmed
+	// this will create an stale entry in conntrack that will blackhole the traffic, so we need to clear it
```
This opens up the opposite race condition: now if someone connects to the service in between when we program iptables and here, that will create a conntrack entry for the DNAT mapping, but then we'll delete that entry here, which means the reply packets will get lost.

Do we _only_ want to delete blackhole entries? If so, can we pass an argument to `conntrack` specifying that? (`-u <something>` maybe?)

Maybe since this is UDP we don't care; we're already claiming below that accidentally clearing out the wrong conntrack entries is "very low impact".

aojea replied:
yeah, I was testing this patch and it keeps failing, I think that I'm hitting exactly what you are describing ...

> Do we only want to delete blackhole entries? If so, can we pass an argument to conntrack specifying that? (-u <something> maybe?)

NodePort can receive from any ip and forward to any endpoint , can you elaborate?

danwinship replied:
I mean is there some way we can say "delete entries that say 'this IP doesn't exit' but don't delete entries that say 'this IP _does_ exist'"

aojea replied:
maybe if we can delete only the UNREPLIED?
```
udp      17 28 src=192.168.1.150 dst=192.168.1.144 sport=12345 dport=31123 [UNREPLIED] src=192.168.1.144 dst=192.168.1.150 sport=31123 dport=12345 mark=0 use=1
```

conntrack-tools has an status flag but I´m not able to make it work with that status
```
  -u, --status status           Set status, eg. ASSURED
```

aojea replied:
hmm it is not the reply thing :thinking: ... let me try something

aojea replied:
this is solved now

danwinship commented:
> > I was going to make the same argument for hostPort - and maybe I will.
> 
> Do we limit hostports to a specific range?

It appears that we do not limit hostports, but we _do_ prevent you from getting a hostport if it can't open the listening socket. If we removed the port opener check, then you could use hostport forwarding to steal traffic from arbitrary system services.

aojea commented:
> > > I was going to make the same argument for hostPort - and maybe I will.
> > 
> > 
> > Do we limit hostports to a specific range?
> 
> It appears that we do not limit hostports, but we _do_ prevent you from getting a hostport if it can't open the listening socket. If we removed the port opener check, then you could use hostport forwarding to steal traffic from arbitrary system services.

hostPorts in k/k are going away with kubenet and dockershim :smile:

danwinship commented:
> hostPorts in k/k are going away with kubenet and dockershim

well, cri-o at least has its own hostport implementation that works the same way, but I guess in any cluster using the CNI portmap plugin to implement their hostports, users can probably steal arbitrary ports

aojea commented:
> well, cri-o at least has its own hostport implementation that works the same way, but I guess in any cluster using the CNI portmap plugin to implement their hostports, users can probably steal arbitrary ports

yeah, but that falls into the CRI/CNI realm now, there is nothing that Kubernetes can do to avoid it, or do you think kubelet should do the check for hostPorts and coordinate with the CRI?

aojea commented:
/hold cancel

@danwinship @smarterclayton @thockin this is working , I run the test successfully 150 times in a row

aojea commented:
/hold
until @hanlins PR https://github.com/kubernetes/kubernetes/pull/97824 merge with the fix to the kube-proxy unit tests
Full credit to him, I just want to have the PR ready for review https://github.com/kubernetes/kubernetes/pull/98305/commits/9c80b6922361766d26aa2c71946d0ff3904c5b02

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -843,6 +845,11 @@ func (proxier *Proxier) syncProxyRules() {
+			nodePort := svcInfo.NodePort()
+			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
+				klog.V(2).Infof("Stale %s service NodePort %v -> %d", strings.ToLower(string(svcInfo.Protocol())), svcPortName, svcInfo.NodePort())
+				staleNodePorts.Insert(nodePort)
```
It looks like this handles the case of the entire Service going away, but does it handle the case of a single endpoint going away / changing?

The "staleness" code is very confusing...

aojea replied:
yeah, I had to look it several times in the code to understand it, because my understanding was exactly what you are saying.However, the observation during debugging the problem is that the service ClusterIP was declared "stale" when the endpoint was updated but the service not, that is exactly the condition that we are looking for. The test added shows perfectly the behaviour.

I don´t know if this was on purpose or a side effect of how the endpoints/services tracker/map/cache works, but this is the kind of "staleness" that kube-proxy was  using always ... and this is exactly the place where we should delete the nodeport, that is also coherent with the rest of conntrack clearing

aojea replied:
aha, is already documented

https://github.com/kubernetes/kubernetes/blob/074a5177209c3605daf11cfe6f477fddce0d55c0/pkg/proxy/endpoints.go#L534-L536

aojea replied:
added an additional comment, but we must delete the nodeport conntrack entries when endpoints move from 0 to non zero, not when the service is added without endpoints.
This should be ready

aojea commented:
/hold cancel
The depending PR is approved, once it merge this can go in

aojea commented:
/retest

aojea commented:
/retest

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -835,15 +836,24 @@ func (proxier *Proxier) syncProxyRules() {
 	serviceUpdateResult := proxier.serviceMap.Update(proxier.serviceChanges)
 	endpointUpdateResult := proxier.endpointsMap.Update(proxier.endpointsChanges)
 
+	// We need to detect stale Services for UDP so we can clean
+	// dangling conntrack entries that can blackhole traffic.
 	staleServices := serviceUpdateResult.UDPStaleClusterIP
+	staleNodePorts := sets.NewInt()
 	// merge stale services gathered from updateEndpointsMap
+	// an UDP service that changes from 0 to non-0 endpoints is considered stale.
 	for _, svcPortName := range endpointUpdateResult.StaleServiceNames {
```
OK... the reason this needs so many comments is that the (pre-existing) naming is terrible.

`serviceUpdateResult.UDPStaleClusterIP` is ok; that's "UDP Services that used to exist but don't any more, so any references to those Cluster IPs are stale". And assigning that to a variable called `staleServices` therefore makes sense.

But `endpointUpdateResult.StaleServiceNames` is "UDP Services that used to have 0 endpoints but now have more than 0 endpoints". `StaleServiceNames` doesn't explain that at all; the connection to staleness is that there might be stale conntrack entries for the service, not that the service itself is in any way stale... So adding those services to `staleServices` is confusing... (although it's correct, because `staleServices` actually just means "services whose conntrack entries we will delete later").

And then `staleNodePorts` means "newly-added UDP node ports" and is even more confusing...

So... this would be much clearer if you renamed `staleServices` to `staleConntrackServices` or `conntrackCleanupServices` or something and likewise with `staleNodePorts`. The use of "stale" in the log messages below could be made clearer as well.

And `StaleServiceNames` ought to be `NewlyActiveUDPServices` or something like that but maybe not in this PR

aojea replied:
done

danwinship commented:
/lgtm
/approve

aojea commented:
/retest

aojea commented:
/retest
[sig-storage] PersistentVolumes-local [Volume type: dir-link

aojea commented:
/retest

Many flakes lately ....

wojtek-t commented:
@aojea - great finding - thanks a lot for debugging and fixing this!

ialidzhikov commented:
@aojea , does it makes sense to have cherry-picks for this PR?

aojea commented:
> @aojea , does it makes sense to have cherry-picks for this PR?

yes it does, are you volunteering :) ? 
Otherwise I will try to do it next week

aojea commented:
@ialidzhikov FYI https://github.com/kubernetes/kubernetes/pull/99017

ialidzhikov commented:
> > @aojea , does it makes sense to have cherry-picks for this PR?
> 
> yes it does, are you volunteering :) ?
> Otherwise I will try to do it next week

Thanks for asking @aojea , but I don't think I have the required expertise in this area. I just wanted to make sure that we don't miss to backport a PR that deserver a backport. :)
Thank you for taking care!
